src/ray/core_worker/task_manager.cc

```diff
@@ -137,11 +131,33 @@ bool ObjectRefStream::InsertToStream(const ObjectID &object_id, int64_t item_ind
            "feature using `max_retries=0` (task) or `max_task_retries=0` (actor).";
   }
   item_index_to_refs_.emplace(item_index, object_id);
+  max_index_seen_ = std::max(max_index_seen_, item_index);
   return true;
 }
 
-void ObjectRefStream::MarkEndOfStream(int64_t item_index) {
-  end_of_stream_index_ = item_index;
+void ObjectRefStream::MarkEndOfStream(int64_t item_index,
+                                      ObjectID *object_id_in_last_index) {
+  if (end_of_stream_index_ != -1) {
+    return;
+  }
+  // ObjectRefStream should guarantee the max_index_seen_
+  // will always have an object reference to avoid hang.
+  // That said, if there was already an index that's bigger than a given
+  // end of stream index, we should mark that as the end of stream.
+  // It can happen when a task is retried and return less values
+  // (e.g., the second retry is failed by an exception or worker failure).
+  end_of_stream_index_ = std::max(max_index_seen_ + 1, item_index);
+
+  auto end_of_stream_id = GetObjectRefAtIndex(end_of_stream_index_);
+  item_index_to_refs_.emplace(end_of_stream_index_, end_of_stream_id);
+  *object_id_in_last_index = end_of_stream_id;
+  return;
+}
+
+ObjectID ObjectRefStream::GetObjectRefAtIndex(int64_t generator_index) const {
+  RAY_CHECK(generator_index < 100 * 1000 * 1000);
+  // 2 for a single return from a generator task.
```

Comment:
// index 1 is reserved for the first task return from a generator task itself.

Copied from `TaskSpecification::StreamingGeneratorReturnId`